### ISSUE TITLE
Add Windows (MSVC) support

### ISSUE THREAD

**Issue by bchretien** [roboptim/roboptim-core#76]
Some people are interested in using RobOptim on Windows, so this is worth investigating.

@kayusawa could you make a summary of your tests regarding the compilation of RobOptim on Windows? Thanks! :-)

**Comment by bchretien** [comment]
Note regarding CI support for Windows: I recently noticed that some open-source projects now use [AppVeyor](https://ci.appveyor.com), which could very roughly be seen as Travis for Windows.

**Comment by phuicy** [comment]
In the latest stable there seems alot of issues with a windows build, that aren't dependencies.

**Comment by bchretien** [comment]
Windows support is not available yet, although we are slowly getting there (cc @aescande).

**Comment by phuicy** [comment]
We have worked out quite few issues, but are still working on alot templating issues:

```
40>C:\git\roboptim-core\include\roboptim/core/problem.hxx(118): error C2491: 'roboptim::Problem<T>::function' : definition of dllimport function not allowed
```
Will keep you informed. Is there an old binary of an older commit?

**Comment by bchretien** [comment]
Thanks for looking into this. Alas I don't think anyone shared any binary yet.

**Comment by francois-keith** [comment]
Just putting my 2 cents here.

I have tried to compile roboptim-core on windows very recently (I must be 1-2 commits late wrt master). Here are the errors I found (as I remember them, so this is not as precise as I'd like it to  be).
* The error you're describing `definition of dllimport function not allowed` is due to the fact that some methods / classes are now templated (but weren't in the last version tested). The error can be suppressed by removing this macro for those classes.
* Some errors will also pop up because of the macro ROBOPTIM_CORE_***_DEPRECATED, that must be placed *before* the function declaration in windows.
* Some of the unit tests will not compile in VS2008 because of template issues (I cannot tell you which ones right here).
* The format of the output is not the same as in unix (e.g., infinity is not written `inf`, but `#INF`, and the E notation takes an extra 0: `1e03` versus `1e003`), so some unit tests may fail.

Those are some leads.
I will push that on my fork (it is still a work in progress) tonight if you want to have a look.

**Comment by phuicy** [comment]
Awesome, this is exactly what we are seeing. 
I think to solve the 1st issue the ROBOPTIM_DLLAPI has to moved to a pre-declaration of the class., as:
```
// function definition
void __declspec(dllimport) funcB() {}   // C2491

// function declaration
void __declspec(dllimport) funcB();   // OK
```
number 2:
We set
```
#define ROBOPTIM_CORE_**_DEPRECATED 
```

**Comment by francois-keith** [comment]
I don't think it is required to keep the ROBOPTIM_DLLAPI macro for the templated class (e.g. Problem) since the definition of the method is in the .hxx and those files are installed as well.
So there is no point AFAIK in exporting the symbols.

I compiled the unit tests as well as roboptim/roboptim-core-plugin-eigen without having any symbol issue.

**Comment by bchretien** [comment]
Indeed, I'll clean this up a bit. I recently added the possibility to precompile most of the templated classes, since they are now only templated on the matrix traits (dense/sparse). However this is disabled by default since explicit template instantiation is a C++11 feature. I guess the error will appear in some other classes.

**Comment by bchretien** [comment]
`ROBOPTIM_CORE_DEPRECATED` (which is equivalent to `__declspec (deprecated)` on Windows with MSVC) has to be at the beginning of function declarations/typedefs?

**Comment by francois-keith** [comment]
Yes, otherwise some compilation issues may raise.

**Comment by bchretien** [comment]
Ok, we can easily fix that then. I was also considering making a PR to jrl-cmakemodules to have the deprecated macros support version numbers, e.g.:

```cpp
#define ROBOPTIM_CORE_DEPRECATED(X) __attribute__ ((deprecated))
...
typedef Foo Bar ROBOPTIM_CORE_DEPRECATED(3.2);
```

They use this for Gazebo, that way you can see exactly when a given function was deprecated when you get the warning while compiling.

**Comment by bchretien** [comment]
And talking about Gazebo, they had the exact same problem:

```cpp
#if defined(__GNUC__)
#define GAZEBO_DEPRECATED(version) __attribute__((deprecated))
#define GAZEBO_FORCEINLINE __attribute__((always_inline))
#elif defined(_WIN32)
// GAZEBO_DEPRECATED should be defined as something like
// __declspec(deprecated), but it needs to go *before* the function name,
// and we're putting GAZEBO_DEPRECATED *after* the function.
#define GAZEBO_DEPRECATED(version)
#define GAZEBO_FORCEINLINE __forceinline
#else
#define GAZEBO_DEPRECATED(version) ()
#define GAZEBO_FORCEINLINE
#endif
```

**Comment by phuicy** [comment]
Bear in mind, getting Gazebo is still not an easy experience. But this looks good. 
That commit (992e734) looks very similar to what we got.

**Comment by francois-keith** [comment]
@phuicy I pushed on https://github.com/francois-keith/roboptim-core the current state of my windows version.
I tested it on MSVC 2012, with the following deps:
* boost 1.55
* libtools 1.5.26
* eigen 3.2.0

@bchretien No Pull request for now, mainly because of the test issue and the fact that the result file is different between the Win32 and UNIX systems.

**Comment by bchretien** [comment]
@francois-keith good catch on the bug recently added to `boost.cmake` (see jrl-umi3218/jrl-cmakemodules#76). As for the differences when printing, there are some ideas [there](http://stackoverflow.com/questions/9226400/portable-printing-of-exponent-of-a-double-to-c-iostreams). The `_set_output_format` trick could be added to our test fixtures, thus it would be automatically applied to all the RobOptim tests (core, plugins...) by simply changing [a single file](https://github.com/roboptim/roboptim-shared-tests/blob/15950fb8dbd1523645a2cd54718db832dbb5409e/fixture.hh#L39-L56).

**Comment by bchretien** [comment]
@francois-keith well apparently [_set_output_format](https://msdn.microsoft.com/en-us/library/0fatw238.aspx), still any alternative workaround should probably be applied in our test fixtures.

**Comment by bchretien** [comment]
Complementary information [here](https://connect.microsoft.com/VisualStudio/feedback/details/1368280):

> The removal of _set_output_format is by design. The standards-conforming two-digit exponent mode is now the default and the non-conforming, previously-default, three-digit exponent format is no longer supported. This is noted in the draft Visual C++ 2015 breaking changes documentation: https://msdn.microsoft.com/en-us/library/bb531344(v=vs.140).aspx. We have not yet completed the full updates to the documentation for Visual Studio 2015.

So we can just check the Visual C++ version, and use `_set_output_format` if an old version is detected.

**Comment by phuicy** [comment]
Awesome. Thank you

**Comment by bchretien** [comment]
Does anyone know how MSVC behaves with explicit template instantiations and symbol exports? The weird symbol exports I had all around the code solved [this problem](https://llvm.org/bugs/show_bug.cgi?id=24815), i.e. gcc and clang behave differently (either one returns a `type attributes ignored after type is already defined` error, or the other one does not export the symbols).

**Comment by francois-keith** [comment]
AFAIK, you can export the specialization of templated methods or methods of templated class.
I've put an example of the syntax in this repo : francois-keith/test_template_export@34e2e12

To be honest, it was much more tricky in my memory (cf stack-of-tasks/dynamic-graph@5721ccb98), so maybe I'm missing the point / the difficulty here.

**Comment by bchretien** [comment]
The problem is that gcc and clang behave quite differently, so I was wondering if msvc exhibited the same problem.

* The following version works (at least the export) in clang but not gcc:

```cpp
// In the header
template <typename T>
class Problem { ... };
...
extern template class ROBOPTIM_CORE_DLLAPI Problem<EigenMatrixDense>;
extern template class ROBOPTIM_CORE_DLLAPI Problem<EigenMatrixSparse>;

// In the implementation cpp
template class ROBOPTIM_CORE_DLLEXPORT Problem<EigenMatrixDense>;
template class ROBOPTIM_CORE_DLLEXPORT Problem<EigenMatrixSparse>;
```

With clang and the explicit template instantiation:
```sh
$ nm -CD src/libroboptim-core.so | grep "W roboptim::Problem<roboptim::EigenMatrixDense>::initialize()"
00000000000f5170 W roboptim::Problem<roboptim::EigenMatrixDense>::initialize()
```
Without it, the symbol is not exported as expected.

gcc error:
```txt
roboptim-core/include/roboptim/core/problem.hxx:656:46: error: type attributes ignored after type is already defined [-Werror=attributes]
   extern template class ROBOPTIM_CORE_DLLAPI Problem<EigenMatrixSparse>;
                                              ^
```

* The following versions works with both gcc and clang (at least the export), but will fail with msvc:

```cpp
// In the header
template <typename T>
class ROBOPTIM_CORE_DLLAPI Problem { ... };
...
extern template class Problem<EigenMatrixDense>;
extern template class Problem<EigenMatrixSparse>;

// In the implementation cpp
template class ROBOPTIM_CORE_DLLEXPORT Problem<EigenMatrixDense>;
template class ROBOPTIM_CORE_DLLEXPORT Problem<EigenMatrixSparse>;
```

With gcc we finally get:
```sh
$ nm -CD src/libroboptim-core.so | grep "W roboptim::Problem<roboptim::EigenMatrixDense>::initialize()"
00000000000f9e10 W roboptim::Problem<roboptim::EigenMatrixDense>::initialize()
```

**Comment by bchretien** [comment]
Experimental Windows support is currently available on the `dev` branch. The next step would be to validate the changes with a solver plugin.
